File: protocol-generator/dart/test/tree_ladder_test.dart
import 'dart:typed_data';

import 'package:entity_core_protocol/entity_core_peer.dart';
import 'package:entity_core_protocol/src/codec/ecf_value.dart';
import 'package:entity_core_protocol/src/peer/wire.dart' as wire;
import 'package:test/test.dart';

/// §3.3's effective-targets ladder, the §6.3 path check it gates, and the §6.3
/// listing filter (RULE A, 0.8.2.20/.21/.22 + N7/N10/N11 at .24/.25).
///
/// EVERY LADDER CASE GOES THROUGH THE REAL DISPATCH CHAIN with a fully signed
/// envelope, not through the tree handler directly. That is not ceremony: it is what
/// makes the RULE G ordering claim an ordering claim at all, since a test that
/// resolves the operation itself cannot see a handler that validates the resource
/// first. It also means the §6.3 check runs on the authority the DISPATCH check
/// resolved, which is the property §6.8 names.
void main() {
  Uint8List seed(int b) => Uint8List(32)..fillRange(0, 32, b);

  late Peer peer;
  late String local;

  setUp(() async {
    // The discovery floor, NOT open grants: the §6.3 path check is only observable
    // where the caller's grant is narrower than the request.
    peer = await Peer.create(seed(0x5a), openGrants: false, conformance: false);
    local = peer.localPeer;
    peer.store.bind('/$local/app/a', Entity.make('primitive/any', cmap(['v', 1])));
    peer.store.bind('/$local/app/b', Entity.make('primitive/any', cmap(['v', 2])));
  });

  EcfMap scopeOf(List<String> incl, [List<String> excl = const []]) => EcfMap.of({
        'include': textArray(incl),
        if (excl.isNotEmpty) 'exclude': textArray(excl),
      });

  /// A SELF-ISSUED ROOT capability whose three dimensions move independently.
  ///
  /// granter == grantee == this peer's identity hash, which is what §5.5's root arm
  /// requires ("root must be self-issued by the local peer") and what makes the token
  /// usable through the real dispatch chain.
  Entity token({
    List<String> handlers = const ['*'],
    List<String> operations = const ['*'],
    List<String> resources = const ['*'],
    List<String> resourcesExcl = const [],
  }) {
    final idh = peer.identity.identityHash();
    return Entity.make(
      'system/capability/token',
      cmap([
        'granter', cbytes(idh),
        'grantee', cbytes(idh),
        'created_at', EcfInt.of(1700000000000),
        'grants', EcfArray([
          EcfMap.of({
            'handlers': scopeOf(handlers),
            'operations': scopeOf(operations),
            'resources': scopeOf(resources, resourcesExcl),
          }),
        ]),
      ]),
    );
  }

  EcfMap? resourceOf(List<String>? targets, [List<String> excl = const []]) =>
      targets == null
          ? null
          : EcfMap.of({
              'targets': textArray(targets),
              if (excl.isNotEmpty) 'exclude': textArray(excl),
            });

  /// Drive one request through `Peer.dispatch`. Returns (status, code, result).
  Future<(int, String, Entity?)> drive(
    String operation,
    EcfMap? resource, {
    Entity? cap,
    Entity? params,
    bool withCap = true,
  }) async {
    final id = peer.identity;
    final t = cap ?? token();
    final exec = wire.makeExecute(
      'r1',
      'system/tree',
      operation,
      params ?? wire.emptyParams(),
      author: withCap ? id.identityHash() : null,
      capability: withCap ? t.hash() : null,
      resource: resource,
    );
    // §5.2 demands a system/signature over the EXECUTE whose signer binds to
    // `author`, and a chain whose root is self-issued by this peer. Anything less is
    // refused at the verdict and every ladder assertion below would be reading a 401
    // rather than the handler.
    final included = withCap
        ? [t, id.peerEntity, await id.sign(t), await id.sign(exec)]
        : <Entity>[];
    final env = Envelope(
        exec, [for (final e in included) Included(e.rawHash, e)]);
    final resp = await peer.dispatch(Conn(), env);
    expect(resp, isNotNull, reason: 'every inbound root is answered (N12/N17)');
    final status = wire.responseStatus(resp!);
    final res = wire.responseResult(resp);
    return (status, res?.text('code') ?? '', res);
  }

  group('effective_targets: the two-empties discriminator (N11, 0.8.2.25)', () {
    // N11 makes the discriminator a [MUST]: "that projection MUST NOT be lossy about
    // its own emptiness — narrow when narrowing leaves something, and retain the raw
    // pair when narrowing would empty it." This peer carries it as a two-field record
    // rather than a nullable list, so the collapse is not one `?? const []` away.
    EffectiveTargets eff(EcfMap? r) => effectiveTargets(
        local,
        wire.makeExecute(
            'r1', 'system/tree', 'get', wire.emptyParams(), resource: r));

    test('an ABSENT resource reports absent', () {
      expect(eff(null).hadResource, isFalse);
    });

    test('a resource map with no `targets` key reports absent (an OPEN question)', () {
      // PINS THE SHIPPED ANSWER rather than endorsing it. Every 0.8.2.25 peer reports
      // this absent, so `get` serves it the root listing; §3.2 says `targets` "MUST
      // contain at least one entry", which makes the shape MALFORMED rather than
      // absent, and N10's point is that a PRESENT `resource` must not be served the
      // wider absent-case answer. Nothing in the 778-check set drives it and no
      // disposition is pinned, so the behaviour is HELD rather than changed; this
      // case exists so that changing it is a DECISION.
      expect(eff(EcfMap.of({'exclude': textArray(['a'])})).hadResource, isFalse);
    });

    test('a PRESENT but fully-excluded resource is present with an empty list', () {
      final e = eff(resourceOf(['a'], ['a']));
      expect(e.hadResource, isTrue);
      expect(e.survivors, isEmpty);
    });

    test("survivors keep the caller's OWN SPELLING", () {
      // 0.8.2.21: `effective_targets` yields RAW survivors, not canonical forms — the
      // value flows on to the store lookup, which canonicalizes for itself.
      expect(eff(resourceOf(['app/x', 'app/y'], ['app/y'])).survivors,
          equals(['app/x']));
    });

    test('an unmatchable CALLER exclude is fail-OPEN: the target survives', () {
      // §5.4's table rules the caller-exclude arm SEPARATELY from the grant arm and in
      // the opposite direction: canonicalize answers the sentinel, matchesPattern
      // answers false, and the target survives. INHERITED from the primitives here
      // rather than restated — restating it is how the two readings drift.
      expect(eff(resourceOf(['app/a'], ['../nope'])).survivors, equals(['app/a']));
      // The accept-side control: without it a peer that ignored `exclude` entirely
      // would pass the row above for the wrong reason.
      expect(eff(resourceOf(['app/a'], ['app/a'])).survivors, isEmpty);
    });
  });

  group('RULE G: operation resolution precedes resource validation', () {
    // "Resolve the operation first; only then run the §3.3 ladder." A peer that
    // validates the resource first answers a RESOURCE fault for an OPERATION fault on
    // every unknown operation — measured independently by `entity-system-conformance`
    // (X9 / F52) on the peers whose tree handler is one match over
    // (operation, resource).
    //
    // ON THIS SUBSTRATE THE ORDERING IS STRUCTURAL AND THAT IS THE FINDING, not a
    // fix: `_OpsHandler.handle` looks the operation up in a map and answers 501 on a
    // miss, and the §3.3 ladder lives inside `_treeGet`/`_treePut`, which only a known
    // operation reaches. The pair is asserted anyway because "it cannot happen" is a
    // claim about today's dispatcher, and this is what keeps it one.

    test('an unknown operation is 501 WITH and WITHOUT a resource', () async {
      final without = await drive('bogusop', null);
      expect(without.$1, equals(501));
      expect(without.$2, equals('unsupported_operation'));
      // THE CONTROL that makes this an ORDERING claim rather than a missing-501
      // claim: on the peers that had the defect, these two answered DIFFERENTLY.
      final with_ = await drive('bogusop', resourceOf(['app/a']));
      expect(with_.$1, equals(501));
      expect(with_.$2, equals('unsupported_operation'));
      // ...and a KNOWN operation still routes, or the two above are satisfied by a
      // handler that answers 501 to everything.
      expect((await drive('get', resourceOf(['app/a']))).$1, equals(200));
    });
  });

  group('the ladder, get (resource-OPTIONAL, BROAD-RESULT)', () {
    // EXTENSION-TREE §2.2a (v4.11) declares `get` resource-OPTIONAL and BROAD-RESULT:
    // absent-case answer "the root listing", self-excluded case "400 path_required".
    // 0.8.2.24 (N7) scopes §3.3's "an empty effective list IS the absent case" to "an
    // operation that REQUIRES a resource", and 0.8.2.25 (N10) decides the
    // present-but-empty case by whether the absent case is WIDER than the request.

    test('an ABSENT resource serves the root listing', () async {
      final (status, _, res) = await drive('get', null);
      expect(status, equals(200));
      expect(res!.type, equals('system/tree/listing'));
      expect(res.text('path'), equals('/$local/'));
    });

    test('a PRESENT but self-excluded resource is 400 path_required', () async {
      // THE TWO EMPTIES ARE DISTINCT, and this is the pair that says so. Collapsing
      // them serves the ROOT LISTING to a request that named one excluded path — the
      // wider-than-the-request answer §3.3 forbids.
      final (status, code, _) = await drive('get', resourceOf(['app/a'], ['app/a']));
      expect(status, equals(400));
      expect(code, equals('path_required'));
    });

    test('more than one EFFECTIVE target is 400 ambiguous_resource', () async {
      final (status, code, _) = await drive('get', resourceOf(['app/a', 'app/b']));
      expect(status, equals(400));
      expect(code, equals('ambiguous_resource'));
      // ...and the exclude is what makes the COUNT an effective-set count rather than
      // a raw `targets` count: two targets, one excluded, ONE survivor -> it proceeds.
      expect((await drive('get', resourceOf(['app/a', 'app/b'], ['app/b']))).$1,
          equals(200));
    });

    test('the selection is the SURVIVOR, never targets[0]', () async {
      // THE MUST 0.8.2.20 NAMES: a handler that counts the effective list and then
      // indexes `targets[0]` has implemented the arithmetic completely and is still
      // reading a path no authorization covered. `targets[0]` is EXCLUDED here and the
      // single survivor is `targets[1]`, so the two readings return DIFFERENT
      // entities.
      final (status, _, res) =
          await drive('get', resourceOf(['app/a', 'app/b'], ['app/a']));
      expect(status, equals(200));
      expect(res!.uint('v')?.toInt(), equals(2), reason: 'the survivor app/b, not app/a');
    });

    test('a PATTERN target is 400 malformed_resource; a trailing slash is a listing',
        () async {
      // 0.8.2.20: a resource-requiring operation takes a CONCRETE path. A trailing "/"
      // is a listing request and is NOT a pattern — only a star makes it one.
      final pat = await drive('get', resourceOf(['app/*']));
      expect(pat.$1, equals(400));
      expect(pat.$2, equals('malformed_resource'));
      final lst = await drive('get', resourceOf(['app/']));
      expect(lst.$1, equals(200));
      expect(lst.$3!.type, equals('system/tree/listing'));
    });
  });

  group('the ladder, put (resource-REQUIRED)', () {
    test('a MISSING put target is 400 path_required, not ambiguous_resource', () async {
      // THE CODE CHANGE 0.8.2.20 FORCED. This branch answered `ambiguous_resource` for
      // a MISSING target, which 0.8.2.20 names as the exact inversion it forbids: the
      // remedies differ — supply a resource is not disambiguate your request — and the
      // code is what selects the remedy.
      final miss = await drive('put', null);
      expect(miss.$1, equals(400));
      expect(miss.$2, equals('path_required'));
      // BOTH empties collapse here, because §2.2a declares `put` resource-REQUIRED:
      expect((await drive('put', resourceOf(['app/a'], ['app/a']))).$2,
          equals('path_required'));
      // ...and more than one survivor is still `ambiguous_resource`, which says the
      // two codes have not simply been swapped.
      expect((await drive('put', resourceOf(['app/a', 'app/b']))).$2,
          equals('ambiguous_resource'));
    });
  });

  group('section 6.3: the handler-level path check (0.8.2.20)', () {
    // THE F84 LAYER. §6.3: "not a secondary check ... the sole enforcement wherever
    // the subject is derived after dispatch."
    //
    // WHICH RUNG ANSWERS THE SCALAR 403 IS BOTH, and that was measured rather than
    // assumed: this peer's dispatch-level `checkResourceScope` requires every
    // non-caller-excluded target to be covered, so for a SCALAR get the handler's
    // derived path is always a path dispatch already saw, and the two rungs are
    // independently sufficient. The arm where §6.3 is the ONLY thing standing is the
    // LISTING, whose entries the dispatch check never sees — measured below, and that
    // is where a plant on the filter bites.

    test('a GET on a path the caller capability excludes is 403', () async {
      final narrow = token(resources: ['*'], resourcesExcl: ['app/b']);
      final denied = await drive('get', resourceOf(['app/b']), cap: narrow);
      expect(denied.$1, equals(403));
      expect(denied.$2, equals('capability_denied'));
      // CONTROL, in the SAME capability: a path the same grant covers is served.
      expect((await drive('get', resourceOf(['app/a']), cap: narrow)).$1, equals(200));
    });

    test('a PUT on an excluded path is 403 and writes nothing', () async {
      final narrow = token(resources: ['*'], resourcesExcl: ['app/b']);
      final e = Entity.make('primitive/any', cmap(['v', 9]));
      final params = Entity.make('primitive/any', cmap(['entity', e.toCbor()]));
      final out =
          await drive('put', resourceOf(['app/b']), cap: narrow, params: params);
      expect(out.$1, equals(403));
      expect(out.$2, equals('capability_denied'));
      // ...and nothing was written: `app/b` still holds what setUp bound. A 403 whose
      // refusal arrives AFTER the store write would satisfy the status assertion on
      // its own, which is why the store is read here.
      expect(peer.store.getAt('/$local/app/b')!.uint('v')?.toInt(), equals(2));
    });

    test('checkPathPermission moves one dimension at a time', () {
      // The unit-level form. A predicate test built only from DENY cases is
      // indistinguishable from one asserting `false == false`, so the ACCEPT case is
      // what validates the fixture; one deny per DIMENSION is what says the predicate
      // checks the dimension rather than merely denying.
      final wide = token();
      expect(checkPathPermission(local, 'get', 'app/a', wide, 'system/tree'), isTrue);
      expect(
          checkPathPermission(local, 'get', 'app/a', token(operations: ['put']),
              'system/tree'),
          isFalse);
      expect(
          checkPathPermission(local, 'get', 'app/a',
              token(handlers: ['system/other']), 'system/tree'),
          isFalse);
      // An empty `resources.include` is a LEGAL grant shape (§5.2: handlers that touch
      // no tree paths) and DENIES every path here, which is what that note says it
      // should — `covered` over an empty include list is false.
      expect(
          checkPathPermission(
              local, 'get', 'app/a', token(resources: const []), 'system/tree'),
          isFalse);
      // canonicalize is TOTAL and answers the sentinel, which matches no grant, so a
      // malformed path falls through to DENY rather than being matched.
      expect(checkPathPermission(local, 'get', '../escape', wide, 'system/tree'),
          isFalse);
    });
  });

  group('section 6.3: the listing filter (0.8.2.21/.22)', () {
    // "Entries for which check_path_permission returns DENY MUST be omitted. The
    // result's `count` field MUST reflect the FILTERED entry count, not the source
    // tree's total count." A count that still reports the source total IS the
    // disclosure the rule exists to prevent, so it is asserted separately from the
    // entry map: a filter that omits the entry and leaves the count names how many
    // bindings the caller was not allowed to see.

    List<String> segmentsOf(Entity listing) {
      final m = listing.mapField('entries')!;
      return m.entries.map((e) => (e.key as EcfText).value).toList()..sort();
    }

    test('the listing omits an entry the caller cap excludes, and count follows',
        () async {
      final narrow = token(resources: ['*'], resourcesExcl: ['app/b']);
      final (status, _, res) = await drive('get', resourceOf(['app/']), cap: narrow);
      expect(status, equals(200));
      expect(segmentsOf(res!), equals(['a']));
      expect(res.uint('count')?.toInt(), equals(1));
    });

    test('a wide capability sees both entries', () async {
      // The differential that attributes the row above to the FILTER rather than to
      // the store or the directory being wrong: same directory, a grant that covers
      // both, both entries.
      final (_, __, res) = await drive('get', resourceOf(['app/']));
      expect(segmentsOf(res!), equals(['a', 'b']));
      expect(res.uint('count')?.toInt(), equals(2));
    });

    test('the filter narrows on an INCLUDE as well as on an exclude', () async {
      // Both rows above narrow with `exclude`, and a filter that only consulted the
      // exclude list would pass both. This one covers the directory and one child and
      // says nothing about the other.
      //
      // SAY WHAT IS NOT ASSERTED HERE. §6.3's "the DIRECTORY itself is deliberately
      // not checked — each ENTRY is the subject" is NOT separately observable through
      // the dispatch chain on this peer: the dispatch-level check already requires the
      // caller's grant to cover the listing TARGET, so a grant covering only `app/a`
      // is refused at §5.2 before any listing is built, and a grant that does cover
      // `app/` cannot distinguish a filter that checks the prefix from one that does
      // not. The property is real and enforced in `_entryVisible` (the per-entry call
      // is on the CHILD path); it is recorded here as not-driven rather than asserted
      // by a case that would pass either way.
      final dirAndA = token(resources: ['app/', 'app/a']);
      final (status, _, res) = await drive('get', resourceOf(['app/']), cap: dirAndA);
      expect(status, equals(200));
      expect(segmentsOf(res!), equals(['a']));
      expect(res.uint('count')?.toInt(), equals(1));
    });
  });
}

File: protocol-generator/typescript/test/tree-ladder.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  Envelope,
  HandlerContext,
  TreeHandler,
  Ecf,
  Entity,
  type ExecuteResponse,
  Peer,
  type PeerSession,
  ResourceTarget,
  Scope,
  GrantEntry,
  SeedPolicy,
  CapabilityToken,
} from "../src/index.js";
import { checkPathPermission, effectiveTargets } from "../src/capability/permissions.js";
import { type EcfValue, ecfPreEncoded } from "../src/codec/ecf-value.js";
import { Execute } from "../src/model/execute.js";

/**
 * §3.3's effective-targets ladder (0.8.2.20, refined at .24/.25) and §6.3's
 * `check_path_permission`.
 *
 * The ladder runs on the EFFECTIVE list, never on `resource.targets`: a handler that
 * counts the effective list and then indexes `targets[0]` has implemented the arithmetic
 * completely and is still reading a path no authorization covered. Before this, every row
 * below answered `400 handler_error` — a refusal for a reason §6.3 does not name, which is
 * not the selection mechanism and reads as conformance on any check that only asks whether
 * something uncovered was served.
 */

const TIMEOUT = 10_000;
const LOCAL = "12D3KooWLocalPeerIdExampleAAAAAAAAAAAAAAAAAAAA";

async function withSession(fn: (s: PeerSession, peerId: string) => Promise<void>): Promise<void> {
  const responder = new Peer({ seedPolicy: SeedPolicy.debugOpen() });
  const initiator = new Peer();
  try {
    const port = await responder.listen(0);
    const session = await initiator.connect("127.0.0.1", port, TIMEOUT);
    await fn(session, responder.localPeerId);
  } finally {
    await initiator.dispose();
    await responder.dispose();
  }
}

const emptyParams = (): Entity => Entity.create("primitive/any", Ecf.emptyMap());

function codeOf(r: ExecuteResponse): string {
  return Ecf.optText(r.result.data, "code") ?? "";
}

const get = (s: PeerSession, r: ResourceTarget | null): Promise<ExecuteResponse> =>
  s.execute("system/tree", "get", emptyParams(), r, TIMEOUT);
const put = (s: PeerSession, r: ResourceTarget | null): Promise<ExecuteResponse> =>
  s.execute("system/tree", "put", emptyParams(), r, TIMEOUT);

test("§3.3 — the ladder answers each row with its own code", async () => {
  await withSession(async (s) => {
    // `get`, ABSENT resource -> the root listing. EXTENSION-TREE §2.2a (v4.11) declares
    // `get` resource-OPTIONAL and BROAD-RESULT, absent-case answer "the root listing".
    // This used to be a `targets.length !== 1` THROW -> 400 handler_error.
    const absent = await get(s, null);
    assert.equal(absent.statusCode, 200, "an absent resource is the root listing, not a refusal");
    assert.equal(absent.result.type, "system/tree/listing");

    // `get`, PRESENT and self-excluded -> 400 path_required. THE TWO EMPTIES ARE DISTINCT
    // for a resource-OPTIONAL operation (0.8.2.24 N7 / 0.8.2.25 N10): serving this the
    // absent case "answers a request for one excluded path with a listing of the tree".
    const selfExcluded = await get(s, new ResourceTarget(["app/a"], ["app/a"]));
    assert.equal(selfExcluded.statusCode, 400);
    assert.equal(codeOf(selfExcluded), "path_required");

    // Two survivors -> 400 ambiguous_resource. A peer indexing targets[0] answers 200 and
    // cannot tell the caller it ignored the second.
    const ambiguous = await get(s, new ResourceTarget(["app/a", "app/b"], null));
    assert.equal(ambiguous.statusCode, 400);
    assert.equal(codeOf(ambiguous), "ambiguous_resource");

    // A PATTERN subject -> 400 malformed_resource. A resource-requiring operation takes a
    // CONCRETE path; without this the pattern is looked up as a literal and answers 404,
    // which names the wrong fault.
    const pattern = await get(s, new ResourceTarget(["system/type/*"], null));
    assert.equal(pattern.statusCode, 400);
    assert.equal(codeOf(pattern), "malformed_resource");

    // An unmatchable CALLER exclude carves out NOTHING — the fail-OPEN direction of §5.4's
    // sentinel, and the correct one here (§5.4's table rules the caller arm separately
    // from the grant arm). A 400 would mean the peer raised on a path it cannot
    // canonicalize, i.e. the error return 0.8.2.20 removed.
    const unmatchable = await get(s, new ResourceTarget(["system/type/"], ["../nope"]));
    assert.equal(unmatchable.statusCode, 200, "an unmatchable caller exclude must not refuse the request");

    // `put` collapses the two empties: §2.2a declares it resource-REQUIRED, so both answer
    // `path_required`. Note the code 0.8.2.20 forces — a MISSING target is NOT
    // `ambiguous_resource`; *supply a resource* is not *disambiguate your request*.
    for (const r of [null, new ResourceTarget(["app/a"], ["app/a"])]) {
      const res = await put(s, r);
      assert.equal(res.statusCode, 400);
      assert.equal(codeOf(res), "path_required", `put with ${r === null ? "no resource" : "a self-excluded target"}`);
    }
    const putAmbiguous = await put(s, new ResourceTarget(["app/a", "app/b"], null));
    assert.equal(codeOf(putAmbiguous), "ambiguous_resource");

    // RULE G control: the OPERATION resolves FIRST. An unknown op with no resource answers
    // the OPERATION fault, never the resource one — a handler that validates the resource
    // first answers `path_required`/`ambiguous_resource` here and names the wrong fault for
    // every unknown operation.
    const bogus = await s.execute("system/tree", "bogusop", emptyParams(), null, TIMEOUT);
    assert.equal(bogus.statusCode, 501);
    assert.equal(codeOf(bogus), "unsupported_operation");
  });
});

test("§3.3 — the subject is SELECTED from the effective set, never targets[0]", async () => {
  await withSession(async (s, peerId) => {
    const a = Entity.create("test/a", Ecf.emptyMap());
    const b = Entity.create("test/b", Ecf.emptyMap());
    for (const [path, e] of [["app/sel/a", a], ["app/sel/b", b]] as const) {
      const req = Entity.create("system/tree/put-request", Ecf.map(["entity", ecfPreEncoded(e.wireBytes)]));
      const ok = await s.execute("system/tree", "put", req, new ResourceTarget([path], null), TIMEOUT);
      assert.equal(ok.statusCode, 200, `seeding ${path}`);
    }
    // targets:[a,b] exclude:[a]. The effective set is {b}, size 1, so the COUNT rule says
    // proceed — and a raw targets[0] selector proceeds on `a`. Both are bound, so a 200
    // naming `a` is a selection defect and nothing else.
    const res = await get(s, new ResourceTarget(["app/sel/a", "app/sel/b"], ["app/sel/a"]));
    assert.equal(res.statusCode, 200);
    assert.equal(res.result.type, "test/b", "targets[0] would have answered test/a");
    assert.equal(peerId.length > 0, true);
  });
});

// ── the projection itself (0.8.2.25 N11) ─────────────────────────────────────

function execWith(resource: EcfValue | null): Execute {
  const fields: [string, unknown][] = [
    ["request_id", Ecf.text("t1")],
    ["uri", Ecf.text("system/tree")],
    ["operation", Ecf.text("get")],
  ];
  if (resource !== null) {
    fields.push(["resource", resource]);
  }
  return new Execute(Entity.create("system/protocol/execute", Ecf.map(...(fields as [string, never][]))));
}

test("N11 — the projection is not lossy about its own emptiness", () => {
  const res = (targets: unknown, exclude?: string[]): EcfValue =>
    exclude === undefined
      ? Ecf.map(["targets", targets as never])
      : Ecf.map(["targets", targets as never], ["exclude", Ecf.array(exclude.map((e) => Ecf.text(e)))]);

  // ABSENT: nothing was asked for.
  assert.deepEqual(effectiveTargets(execWith(null), LOCAL), { survivors: [], hasResource: false });

  // PRESENT with a survivor, in the CALLER'S OWN SPELLING (0.8.2.21 yields RAW survivors).
  assert.deepEqual(effectiveTargets(execWith(res(Ecf.array([Ecf.text("app/a")]))), LOCAL), {
    survivors: ["app/a"],
    hasResource: true,
  });

  // PRESENT and SELF-EXCLUDED: the discriminator N11 makes a MUST. A function returning
  // only a list collapses this into the absent case and the handler's refusal arm becomes
  // dead code that only a WIRE drive can detect.
  assert.deepEqual(effectiveTargets(execWith(res(Ecf.array([Ecf.text("app/a")]), ["app/a"])), LOCAL), {
    survivors: [],
    hasResource: true,
  });

  // A `resource` map with NO `targets` key is ABSENT.
  assert.deepEqual(
    effectiveTargets(execWith(Ecf.map(["exclude", Ecf.array([Ecf.text("app/a")])])), LOCAL).hasResource,
    false,
  );

  // A PRESENT-BUT-ILL-TYPED `targets` is PRESENT with an EMPTY survivor list, never
  // absent. Reporting it absent serves the WIDER absent-case answer to a request that
  // named a resource — N11's own defect one field over, and the cell the two vanguard
  // peers initially disagreed on.
  assert.deepEqual(effectiveTargets(execWith(res(Ecf.uint(42n))), LOCAL), { survivors: [], hasResource: true });

  // The caller-exclude arm is fail-OPEN on an unmatchable pattern (§5.4 rules the caller
  // arm separately from the grant arm), and that is INHERITED from the matcher, not
  // restated: `../nope` canonicalizes to NEVER_MATCH and matches nothing.
  assert.deepEqual(effectiveTargets(execWith(res(Ecf.array([Ecf.text("app/a")]), ["../nope"])), LOCAL).survivors, [
    "app/a",
  ]);
});

// ── §6.3's handler-level path check ──────────────────────────────────────────

test("§6.3 — check_path_permission accepts what a grant covers and denies each dimension", () => {
  const token = (handlers: string[], operations: string[], resources: string[]): CapabilityToken =>
    new CapabilityToken(
      Entity.create(
        "system/capability/token",
        Ecf.map(
          [
            "grants",
            Ecf.array([
              new GrantEntry(
                new Scope(handlers, null),
                new Scope(resources, null),
                new Scope(operations, null),
                null,
                null,
                null,
              ).toEcf(),
            ]),
          ],
          // `granter`/`grantee`/`created_at` are REQUIRED by the §3.6 parser and are not
          // read by `check_path_permission` — it answers about the token's GRANTS only.
          // The chain, signature, temporal bounds and revocation are `verifyRequest`'s and
          // must already have held.
          ["granter", Ecf.bytes(new Uint8Array(33))],
          ["grantee", Ecf.bytes(new Uint8Array(33))],
          ["created_at", Ecf.uint(0n)],
        ),
      ),
    );

  // THE ACCEPT CASE, and it is the one that validates the FIXTURE: with only deny cases a
  // broken fixture (every scope parsing empty) makes all of them pass for free.
  assert.equal(
    checkPathPermission("get", "app/a", token(["system/tree"], ["get"], ["app/a"]), "system/tree", LOCAL),
    true,
  );
  // One deny per DIMENSION: a single deny cannot distinguish "the predicate checks the
  // dimension I care about" from "the predicate denies".
  assert.equal(
    checkPathPermission("put", "app/a", token(["system/tree"], ["get"], ["app/a"]), "system/tree", LOCAL),
    false,
    "operations",
  );
  assert.equal(
    checkPathPermission("get", "app/b", token(["system/tree"], ["get"], ["app/a"]), "system/tree", LOCAL),
    false,
    "resources",
  );
  assert.equal(
    checkPathPermission("get", "app/a", token(["system/other"], ["get"], ["app/a"]), "system/tree", LOCAL),
    false,
    "handlers",
  );
  // An EMPTY `resources.include` is a legal grant shape (§5.2: handlers that touch no tree
  // paths) and denies every path here — `covered` over an empty include list is false.
  assert.equal(
    checkPathPermission("get", "app/a", token(["system/tree"], ["get"], []), "system/tree", LOCAL),
    false,
    "an empty resources.include denies every path",
  );
  // A malformed path canonicalizes to NEVER_MATCH, which matches no grant, so it falls
  // through to DENY rather than escaping as a throw.
  assert.equal(
    checkPathPermission("get", "../nope", token(["system/tree"], ["get"], ["*"]), "system/tree", LOCAL),
    false,
    "a malformed path denies rather than throwing",
  );
});

// ── §6.3's listing filter (0.8.2.21/.22) ─────────────────────────────────────

// "When any handler returns a multi-entry result whose entries are tree paths, each entry
// MUST be individually checked using check_path_permission. Entries for which
// check_path_permission returns DENY MUST be omitted. The result's `count` field MUST
// reflect the filtered entry count, not the source tree's total count."
//
// THIS TEST EXISTS BECAUSE ITS PLANT RAN GREEN WITHOUT IT. Removing the per-entry check
// from the handler left all 140 other cases passing: the filter is measured on the wire by
// `tools/arc-probe` G4, which needs a MINTED narrow capability, and the peer's own gate
// carried nothing. An inert control is not a control.
//
// Driven through a hand-built HandlerContext rather than a session, because the narrow
// grant is the whole input and minting one over the wire would put three more moving parts
// between the assertion and the thing asserted.
test("§6.3 — a listing omits entries the caller's own capability excludes", async () => {
  const peer = new Peer({ seedPolicy: SeedPolicy.debugOpen() });
  try {
    const base = "/" + peer.localPeerId + "/app/list";
    const e = Entity.create("test/leaf", Ecf.emptyMap());
    peer.tree.put(base + "/a", e);
    peer.tree.put(base + "/b", e);

    const narrowToken = (resources: string[]): CapabilityToken =>
      new CapabilityToken(
        Entity.create(
          "system/capability/token",
          Ecf.map(
            [
              "grants",
              Ecf.array([
                new GrantEntry(
                  new Scope(["system/tree"], null),
                  new Scope(resources, null),
                  new Scope(["get"], null),
                  null,
                  null,
                  null,
                ).toEcf(),
              ]),
            ],
            ["granter", Ecf.bytes(new Uint8Array(33))],
            ["grantee", Ecf.bytes(new Uint8Array(33))],
            ["created_at", Ecf.uint(0n)],
          ),
        ),
      );

    const listUnder = async (cap: CapabilityToken | null): Promise<{ names: string[]; count: number }> => {
      const execute = new Execute(
        Entity.create(
          "system/protocol/execute",
          Ecf.map(
            ["request_id", Ecf.text("l1")],
            ["uri", Ecf.text("system/tree")],
            ["operation", Ecf.text("get")],
            ["params", ecfPreEncoded(emptyParams().wireBytes)],
            ["resource", new ResourceTarget(["app/list/"], null).toEcf()],
          ),
        ),
      );
      const ctx = new HandlerContext({
        peer,
        execute,
        envelope: new Envelope(execute.entity, []),
        pattern: "system/tree",
        suffix: "",
        callerCapability: cap,
        handlerGrant: null,
        author: null,
        connection: null,
      });
      const result = await new TreeHandler().handle(ctx);
      assert.equal(result.status, 200);
      const entries = Ecf.require(result.result.data, "entries");
      const names = entries.kind === "map" ? entries.pairs.map(([k]) => (k.kind === "text" ? k.value : "")) : [];
      return { names, count: Number(Ecf.requireUint(result.result.data, "count")) };
    };

    // THE CONTROL, and it is what makes the assertion below falsifiable: with a grant
    // covering BOTH, the listing names both. "b is absent" under a narrower grant is the
    // trivial truth if the directory read does not work at all.
    const wide = await listUnder(narrowToken(["app/list/*"]));
    assert.deepEqual(wide.names.sort(), ["a", "b"], "control: an unfiltered listing names both entries");
    assert.equal(wide.count, 2);

    const narrow = await listUnder(narrowToken(["app/list/a"]));
    assert.deepEqual(narrow.names, ["a"], "an entry the caller's capability excludes MUST be omitted");
    assert.equal(narrow.count, 1, "`count` MUST follow the FILTERED total, not the source tree's");

    // An UNAUTHENTICATED context is not filtered — the filter's subject is "the caller's
    // verified capability", and where there is none there is no caller to narrow. This is
    // the bootstrap path, and it matches both vanguard peers.
    const bootstrap = await listUnder(null);
    assert.deepEqual(bootstrap.names.sort(), ["a", "b"], "an unauthenticated context is not filtered");
  } finally {
    await peer.dispose();
  }
});
